### PR TITLE
gc_worker: fix deadlock in physicalScanAndResolveLocks

### LINKED ISSUE #16304: GC worker blocks forever
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?

Test green gc with 1 TiDB and 4 TiKV.

### 3. What did you see instead?
GC worker blocks forever.
1 goroutine blocks here:
```
 0  0x0000000000437430 in runtime.gopark
    at /usr/lib/go-1.14.1/src/runtime/proc.go:305
 1  0x00000000004052fd in runtime.chansend
    at /usr/lib/go-1.14.1/src/runtime/chan.go:236
 2  0x00000000004050c5 in runtime.chansend1
    at /usr/lib/go-1.14.1/src/runtime/chan.go:127
 3  0x0000000002469201 in github.com/pingcap/tidb/store/tikv/gcworker.(*GCWorker).physicalScanAndResolveLocks
    at /home/youjiali1995/go/pkg/mod/github.com/youjiali1995/tidb@v1.1.0-beta.0.20200411103223-b7c621421015/store/tikv/gcworker/gc_worker.go:1276
 4  0x00000000024652f0 in github.com/pingcap/tidb/store/tikv/gcworker.(*GCWorker).resolveLocksPhysical
    at /home/youjiali1995/go/pkg/mod/github.com/youjiali1995/tidb@v1.1.0-beta.0.20200411103223-b7c621421015/store/tikv/gcworker/gc_worker.go:1055
 5  0x00000000024624aa in github.com/pingcap/tidb/store/tikv/gcworker.(*GCWorker).resolveLocks
    at /home/youjiali1995/go/pkg/mod/github.com/youjiali1995/tidb@v1.1.0-beta.0.20200411103223-b7c621421015/store/tikv/gcworker/gc_worker.go:893
 6  0x000000000246f2e9 in github.com/pingcap/tidb/store/tikv/gcworker.RunDistributedGCJob
    at /home/youjiali1995/go/pkg/mod/github.com/youjiali1995/tidb@v1.1.0-beta.0.20200411103223-b7c621421015/store/tikv/gcworker/gc_worker.go:1702
 7  0x0000000002474160 in github.com/pingcap/tipocket/tests/resolve-lock.(*resolveLockClient).Start
    at ./tests/resolve-lock/resolve_lock.go:224
 8  0x000000000186f08a in github.com/pingcap/tipocket/pkg/control.(*Controller).RunSelfScheduled.func2
    at ./pkg/control/control.go:262
 9  0x0000000000cbd829 in golang.org/x/sync/errgroup.(*Group).Go.func1
    at /home/youjiali1995/go/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:57
10  0x0000000000467091 in runtime.goexit
    at /usr/lib/go-1.14.1/src/runtime/asm_amd64.s:1373
```
And 4 goroutines block here:
```
0  0x0000000000437430 in runtime.gopark
   at /usr/lib/go-1.14.1/src/runtime/proc.go:305
1  0x00000000004052fd in runtime.chansend
   at /usr/lib/go-1.14.1/src/runtime/chan.go:236
2  0x00000000004050c5 in runtime.chansend1
   at /usr/lib/go-1.14.1/src/runtime/chan.go:127
3  0x0000000002470436 in github.com/pingcap/tidb/store/tikv/gcworker.(*mergeLockScanner).physicalScanLocksForStore
   at /home/youjiali1995/go/pkg/mod/github.com/youjiali1995/tidb@v1.1.0-beta.0.20200411103223-b7c621421015/store/tikv/gcworker/gc_worker.go:1950
4  0x0000000002471b7b in github.com/pingcap/tidb/store/tikv/gcworker.(*mergeLockScanner).Start.func1
   at /home/youjiali1995/go/pkg/mod/github.com/youjiali1995/tidb@v1.1.0-beta.0.20200411103223-b7c621421015/store/tikv/gcworker/gc_worker.go:1855
5  0x0000000000467091 in runtime.goexit
   at /usr/lib/go-1.14.1/src/runtime/asm_amd64.s:1373
```

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
4.0

### PR DESCRIPTION
**Pull request by youjiali1995**
Signed-off-by: youjiali1995 <zlwgx1023@gmail.com>

### What problem does this PR solve?

Issue Number: close #16304  

Problem Summary:
It's possible to block on channel sending in physicalScanAndResolveLocks when error occours.

### What is changed and how it works?
What's Changed:
Check `ctx.Done()` when sending to a channel.

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

### Release note 
Fix  the issue that GC worker blocks on channel sending forever when error occours.

### CODE REVIEW

**Comment by sticnarf** [comment]
LGTM. Is there any way we can confirm the issue is resolved?

**Comment by youjiali1995** [comment]
> LGTM. Is there any way we can confirm the issue is resolved?

I added a test.

**Comment by youjiali1995** [comment]
@MyonKeminta PTAL

File: store/tikv/gcworker/gc_worker.go [hunk 1]
```diff
@@ -1256,61 +1256,60 @@ func (w *GCWorker) removeLockObservers(ctx context.Context, safePoint uint64, st
...
-						// Closed
+						// All locks have been resolved.
 						return
 					}
-					err := w.resolveLocksAcrossRegions(innerCtx, locks)
+					err := w.resolveLocksAcrossRegions(ctx, locks)
 					if err != nil {
-						logutil.Logger(innerCtx).Error("resolve locks failed", zap.Error(err))
+						logutil.Logger(ctx).Error("resolve locks failed", zap.Error(err))
 						errCh <- err
```
Comment (MyonKeminta) [inline store/tikv/gcworker/gc_worker.go:1287]: I think here it should check ctx too.

Comment (youjiali1995) [reply store/tikv/gcworker/gc_worker.go:1287]: It's a buffered channel.

**Pull request merged**